clickhouse: 23.3.13.6 -> 23.10.3.5

edef1c opened the pull request:
## Description of changes

New stable version. Supersedes #263026.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

edef1c committed:
clickhouse: 23.3.13.6 -> 23.10.3.5

yu-re-ka commented:
Thanks!

yu-re-ka commented:
CC @domenkozar for the darwin failure

flokli commented on pkgs/servers/clickhouse/default.nix:
```diff
@@ -6,6 +6,7 @@
+, nasm
 , yasm
 , nixosTests
 , darwin
```
This is supposed to go to L15, but GH doesn't let me: `rustSupport` doesn't seem to be for blake3 only anymore, and we should check if it still builds at all without it?

flokli replied:
I updated the comment, and verified it still builds without it.

domenkozar commented:
darwin looks like an upstream issue: https://github.com/ClickHouse/ClickHouse/issues/49892#issuecomment-1549203333

flokli commented:
> Adding `darwin.libpthread` might fix it.

This (and variations of it) was some of the things I tried, as well as dropping the `find_package(Threads REQUIRED)` line, but unfortunately none of it worked. cc @NixOS/darwin-maintainers.

https://github.com/ClickHouse/ClickHouse/issues/49892#issuecomment-1549203333 mentions the same log output, but in a different context (and not even clear if on Darwin).

--

The new version brings a lot of improvements, and I'm not sure how popular running clickhouse on darwin is. I'm tempted to drop Darwin from the list of supported platforms for this PR - it shouldn't block other platforms and people could still revert to an old nixpkgs checkout.

yu-re-ka commented:
I think @domenkozar added darwin support while adding the clickhouse module to devenv: https://github.com/cachix/devenv/pull/660

In that PR it says "Blocked on [clickhouse: build on darwin]"

yu-re-ka commented:
I'm also somewhat of the opinion that we can't maintain the darwin version if there is no maintainer using clickhouse on darwin and actively participating in upgrades etc.

veprbl commented:
> The new version brings a lot of improvements, and I'm not sure how popular running clickhouse on darwin is. I'm tempted to drop Darwin from the list of supported platforms for this PR - it shouldn't block other platforms and people could still revert to an old nixpkgs checkout.

Sounds reasonable. The way to go is `meta.broken = stdenv.isDarwin`, `meta.platforms` is for marking in-principle support.

reckenrode commented:
The failure is because the wrapper assumes you are using lld.ld64, but ClickHouse is not. It’s trying to invoke lld directly, which doesn’t support the ld64-specific arguments in that mode and fails. According to https://github.com/ClickHouse/ClickHouse/commit/1845bfc6446f033475ae18799e20af4daface821, it’s doing that because lld.ld64 generates incompatible unwind information that causes the Debug build of the server to fail to start.

The following patch does two things:
* Suppresses an error due to an implicit integer to floating point conversion; and
* Use lld.ld64 again under the assumption this is not a Debug build (otherwise mark it broken on Darwin).

```patch
diff --git a/pkgs/servers/clickhouse/default.nix b/pkgs/servers/clickhouse/default.nix
index f50037ab8b99..0b23a62ab88c 100644
--- a/pkgs/servers/clickhouse/default.nix
+++ b/pkgs/servers/clickhouse/default.nix
@@ -62,6 +62,15 @@ in mkDerivation rec {
     '';
   };
 
+  patches = [
+    # Suppress -Wimplicit-const-int-float-conversion error in MemoryTracker.cpp.
+    (fetchpatch {
+      url = "https://github.com/ClickHouse/ClickHouse/commit/3da0945316f35c43814301db29d50760a17c5242.patch";
+      hash = "sha256-/qvUcS7/FCUTyO88V8VIAhp/qI/anYRTaowy1TPtT5M=";
+      includes = [ "cmake/warnings.cmake" ];
+    })
+  ];
+
   strictDeps = true;
   nativeBuildInputs = [
     cmake
@@ -152,6 +161,10 @@ in mkDerivation rec {
     popd
 
     cargoSetupPostPatchHook() { true; }
+  '' + lib.optionalString stdenv.isDarwin ''
+    # Make sure Darwin invokes lld.ld64 not lld.
+    substituteInPlace cmake/tools.cmake \
+      --replace '--ld-path=''${LLD_PATH}' '-fuse-ld=lld'
   '';
 
   cmakeFlags = [
```

reckenrode commented:
Note: It’s still building (aarch64-darwin), but I wanted to get that out there since there was an active discussion.

reckenrode commented:
I also had to add `-Wno-unneeded-internal-declaration`, but it built successfully on aarch64-darwin.

flokli commented:
Thanks for the patch, I kicked off some builds, will report back.

The `x86_64-darwin` build still fails very early:


```
@nix { "action": "setPhase", "phase": "buildPhase" }
building
build flags: -j8
[0/2] Re-checking globbed directories...
[1/10015] Building CXX object contrib/libunwind-cmake/CMakeFiles/unwind.dir/__/libunwind/src/Unwind-EHABI.cpp.o
[2/10015] Building CXX object contrib/libunwind-cmake/CMakeFiles/unwind.dir/__/libunwind/src/Unwind-seh.cpp.o
FAILED: contrib/libunwind-cmake/CMakeFiles/unwind.dir/__/libunwind/src/Unwind-seh.cpp.o 
/nix/store/dx25g7vqky7bp48q2v567wr4qgrpdqmf-clang-wrapper-16.0.6/bin/clang++ -D_DEBUG -D_LIBUNWIND_IS_NATIVE_ONLY -D_LIBUNWIND_NO_HEAP=1 -isystem /tmp/nix-build-clickhouse-23.10.3.5.drv-0/source/contrib/libunwind/include -fdiagnostics-color=always -fsized-deallocation  -mpopcnt -fasynchronous-unwind-tables -ffile-prefix-map=/tmp/nix-build-clickhouse-23.10.3.5.drv-0/source=. -falign-functions=32 -mbranches-within-32B-boundaries  -stdlib=libc++ -fdiagnostics-absolute-paths -fstrict-vtable-pointers -O3 -DNDEBUG -std=c++2b -mmacosx-version-min=10.15   -D OS_DARWIN -O3 -fno-exceptions -funwind-tables -fno-sanitize=all -nostdinc++ -fno-rtti -MD -MT contrib/libunwind-cmake/CMakeFiles/unwind.dir/__/libunwind/src/Unwind-seh.cpp.o -MF contrib/libunwind-cmake/CMakeFiles/unwind.dir/__/libunwind/src/Unwind-seh.cpp.o.d -o contrib/libunwind-cmake/CMakeFiles/unwind.dir/__/libunwind/src/Unwind-seh.cpp.o -c /tmp/nix-build-clickhouse-23.10.3.5.drv-0/source/contrib/libunwind/src/Unwind-seh.cpp
In file included from /tmp/nix-build-clickhouse-23.10.3.5.drv-0/source/contrib/libunwind/src/Unwind-seh.cpp:13:
In file included from /tmp/nix-build-clickhouse-23.10.3.5.drv-0/source/contrib/libunwind/src/config.h:16:
In file included from /nix/store/jh0ibzj56bmw6vd9g67jc0x7cv8wmm7m-libSystem-11.0.0/include/assert.h:44:
In file included from /nix/store/jh0ibzj56bmw6vd9g67jc0x7cv8wmm7m-libSystem-11.0.0/include/stdlib.h:128:
/nix/store/jh0ibzj56bmw6vd9g67jc0x7cv8wmm7m-libSystem-11.0.0/include/malloc/_malloc.h:40:48: error: expected function body after function declarator
void    *malloc(size_t __size) __result_use_check __alloc_size(1);
                                                  ^
/nix/store/jh0ibzj56bmw6vd9g67jc0x7cv8wmm7m-libSystem-11.0.0/include/malloc/_malloc.h:41:64: error: expected function body after function declarator
void    *calloc(size_t __count, size_t __size) __result_use_check __alloc_size(1,2);
                                                                  ^
/nix/store/jh0ibzj56bmw6vd9g67jc0x7cv8wmm7m-libSystem-11.0.0/include/malloc/_malloc.h:43:62: error: expected function body after function declarator
void    *realloc(void *__ptr, size_t __size) __result_use_check __alloc_size(2);
                                                                ^
/nix/store/jh0ibzj56bmw6vd9g67jc0x7cv8wmm7m-libSystem-11.0.0/include/malloc/_malloc.h:45:22: error: expected function body after function declarator
void    *valloc(size_t) __alloc_size(1);
                        ^
/nix/store/jh0ibzj56bmw6vd9g67jc0x7cv8wmm7m-libSystem-11.0.0/include/malloc/_malloc.h:50:78: error: expected function body after function declarator
void    *aligned_alloc(size_t __alignment, size_t __size) __result_use_check __alloc_size(2) __OSX_AVAILABLE(10.15) __IOS_AVAILABLE(13.0) __TVOS_AVAILABLE(13.0) __WATCHOS_AVAILABLE(6.0);
                                                                             ^
```

I uploaded the full build log [here](https://gist.github.com/flokli/bc08b59653f2024220298ee46303900c).

reckenrode committed:
clickhose: fix build on aarch64-darwin

flokli committed:
clickhouse: drop comment on rustSupport flag

It's not just blake3 that's provided in there.

Also verified clickhouse still builds even when this is set to false

flokli commented:
I successfully built it on aarch64-darwin as well, thanks @reckenrode! Cherry-picked your patch (and some non-rebuild comment fix).

I guess the only thing left to sort out now is what to do with the broken `x86_64-darwin` build (which is not broken in current nixpkgs)

reckenrode commented:
> I guess the only thing left to sort out now is what to do with the broken `x86_64-darwin` build (which is not broken in current nixpkgs)

If I do `nix build 'github:NixOS/nixpkgs?ref=pull/267033/head#legacyPackages.x86_64-darwin.clickhouse'`, it builds for me. It also builds when I cherry-pick this PR onto staging-next. I don’t know why that build is failing, especially where it is in that log. It looks like the kind of weird failure that happens due to mixing SDK versions, but I ClickHouse isn’t using any frameworks (just the newer SDK version).

veprbl commented:
Erroring build on ofborg is just a timeout.

flokli commented:
> > I guess the only thing left to sort out now is what to do with the broken `x86_64-darwin` build (which is not broken in current nixpkgs)
> 
> If I do `nix build 'github:NixOS/nixpkgs?ref=pull/267033/head#legacyPackages.x86_64-darwin.clickhouse'`, it builds for me. It also builds when I cherry-pick this PR onto staging-next. I don’t know why that build is failing, especially where it is in that log. It looks like the kind of weird failure that happens due to mixing SDK versions, but I ClickHouse isn’t using any frameworks (just the newer SDK version).

Mmh, maybe the way I ran the build (just setting `--argstr system x86_64-linux`), or the way my MacOS system looks like, or something else was the issue. If it builds for you then it's probably fine and I hit a fluke.

Let's get this in and see what hydra says :-)

veprbl commented:
Result of `nixpkgs-review pr 267033` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>clickhouse</li>
  </ul>
</details>
